src/System Application/App/Json/src/JsonImpl.Codeunit.al

```diff
@@ -0,0 +1,494 @@
+    var
+        JProperty: DotNet JProperty;
+        JToken: DotNet JToken;
+    begin
+        Clear(value);
+        if JObject.TryGetValue(propertyName, JToken) then begin
+            JProperty := JObject.Property(propertyName);
+            value := JProperty.Value;
+            exit(true);
+        end;
+    end;
+
+    local procedure ReplaceOrAddJPropertyInJObject(var JObject: DotNet JObject; propertyName: Text; value: Variant): Boolean
+    var
+        JProperty: DotNet JProperty;
+        OldProperty: DotNet JProperty;
+        oldValue: Variant;
+    begin
+        JProperty := JObject.Property(propertyName);
+        if not IsNull(JProperty) then begin
+            OldProperty := JObject.Property(propertyName);
+            oldValue := OldProperty.Value;
+            JProperty.Replace(JProperty.JProperty(propertyName, value));
+            exit(Format(oldValue) <> Format(value));
+        end;
+
+        AddJPropertyToJObject(JObject, propertyName, value);
+        exit(true);
+    end;
+
+    local procedure AddJPropertyToJObject(var JObject: DotNet JObject; propertyName: Text; value: Variant)
+    var
+        JObject2: DotNet JObject;
+        JProperty: DotNet JProperty;
+        ValueText: Text;
+    begin
+        case true of
+            value.IsDotNet:
+                begin
+                    JObject2 := value;
+                    JObject.Add(propertyName, JObject2);
+                end;
+            value.IsInteger,
+            value.IsDecimal,
+            value.IsBoolean:
+                begin
+                    JProperty := JProperty.JProperty(propertyName, value);
+                    JObject.Add(JProperty);
+                end;
+            else begin
+                ValueText := Format(value, 0, 9);
+                JProperty := JProperty.JProperty(propertyName, ValueText);
+                JObject.Add(JProperty);
+            end;
+        end;
+    end;
+
+    local procedure ReplaceJObjectInCollection(var JArray: DotNet JArray; Index: Integer; JSONString: Text): Boolean
+    var
+        JObject: DotNet JObject;
+    begin
+        if not GetJObjectFromCollectionByIndex(JObject, Index) then
+            exit(false);
+
+        if JSONString <> '' then
+            JObject := JObject.Parse(JSONString)
+        else
+            InitializeEmptyObject();
+
+        JArray.RemoveAt(Index);
+        JArray.Insert(Index, JObject);
+        exit(true);
+    end;
+
+    local procedure AddJObjectToCollection(var JArray: DotNet JArray; JSONString: Text): Boolean
+    var
+        JObject: DotNet JObject;
+    begin
+        if JSONString <> '' then
+            JObject := JObject.Parse(JSONString)
+        else
+            InitializeEmptyObject();
+
+        AddJObjectToCollection(JArray, JObject);
+        exit(true);
+    end;
+
+    local procedure AddJObjectToCollection(var JArray: DotNet JArray; JObject: DotNet JObject)
```

Comment:
There are still some issues here. You're not using the local JArray. Instead you're using the global JsonArray. I also get a little confused that we've got a type JsonArray and and a global .NET called JsonArray. Mabye the global variable could be called JsonArrayDotNet - just to emphasize that this is a .NET type and not the language type?

Reply:
@JesperSchulz refactored in [this ](https://github.com/microsoft/BCApps/pull/716/commits/575d69689c464afd9ed79ff538c137b80a0a8a00) commit